util/reggen/gen_tock.py

```diff
@@ -0,0 +1,384 @@
+            if i > 0 and ch == '/' and line[i - 1] == '/':
+                comment = True
+            if not comment and ch in ")]}":
+                indent -= 1
+            if i == 0:
+                indented.append(' ' * indent * amount)
+            if not comment and ch in "([{":
+                indent += 1
+            indented.append(ch)
+        result.append(''.join(indented))
+        if line == '':
+            result.append('')
+    return '\n'.join(result)
+
+
+def to_snake_case(s: str) -> str:
+    val = []
+    for i, ch in enumerate(s):
+        if i > 0 and ch.isupper():
+            val.append('_')
+        val.append(ch)
+    return ''.join(val)
+
+
+def to_studly_caps(s: str) -> str:
+    words = [w.capitalize() for w in s.split('_')]
+    return ''.join(words)
+
+
+def to_upper_snake_case(s: str) -> str:
+    s = s.upper()
+    r = ''
+    for i in range(0, len(s)):
+        r += s[i] if s[i].isalnum() else '_'
+    return r
+
+
+def first_line(s: str) -> str:
+    """Returns the first line of a multi-line string"""
+    return s.splitlines()[0]
+
+
+def format_comment(s: str) -> str:
+    """Formats a string to comment wrapped to a 100 character line width
+
+    Returns wrapped string including newline and // comment characters.
+    """
+    comment = textwrap.wrap(s,
+                            width=97,
+                            initial_indent='// ',
+                            subsequent_indent='// ')
+    return '\n'.join(comment) + '\n'
+
+
+def data_type(name: str, val: int, as_hex: bool) -> str:
+    """ Returns proper data type for name-value pair. """
+
+    if name.endswith("_OFFSET") or name.endswith("_BASE_ADDR"):
+        return "usize"
+
+    if val.bit_length() > 32:
+        log.error(name + " value exceeds 32 bit " + str(val))
+        sys.exit(1)
+
+    if not as_hex and val < 0:
+        return "i32"
+
+    return "u32"
+
+
+def gen_const(outstr: TextIO,
+              name: str,
+              suffix: str,
+              val: int,
+              existing_defines: Set[str],
+              as_hex: bool = False) -> str:
+    r"""Produces a pub const string. Result includes newline.
+
+    Arguments:
+    name - Name of the constant
+    val - Value of the constant
+    existing_defines - set of already generated define names.
+        Error if `name` is in `existing_defines`.
+
+    Example result:
+    name = 'A_NAME'
+    val = '10'
+
+    pub const A_NAME: u32 = 10
+
+    """
+
+    suffix = '' if not suffix.strip() else '_' + suffix
+    name = name + suffix
+    if name in existing_defines:
+        log.error("Duplicate pub const for " + name)
+        sys.exit(1)
+
+    define_declare = 'pub const ' + name + ': ' + data_type(name, val, as_hex)
+
+    val_str = hex(val) if as_hex else str(val)
+    oneline_define = define_declare + ' = ' + val_str + ';'
+
+    existing_defines.add(name)
+
+    output = oneline_define + '\n'
+    genout(outstr, output)
+    return output
```

Comment:
Same here, is this a [duplicate](https://github.com/lowRISC/opentitan/blob/ec30c9fbee87177654c3c5d0999c119875849d73/util/reggen/gen_rust.py#L79-L116)?

Reply:
I assume this just applies throughout.